### PR TITLE
Bug 531949 - Support JSON-LD & Jena 3.6.0 switch

### LINKED ISSUE #11: Migrate to Jena 3.6.0
should we migrate Core to use the Jena 3.6.0?
All CQs are now already approved
https://docs.google.com/spreadsheets/d/1L0pFrUTW7c4Irt8eti6riersSZF3jNl_hXxFrrZBq84/edit?ts=5af0d1d6#gid=242671613

### PR DESCRIPTION
**Pull request by berezovskyi**
Initial support for JSON-LD. Almost didn't use anything from @neormx
patch due to high amount of duplication in his code.

Closes #11

### CODE REVIEW

**Comment by eclipsewebmaster** [comment]
Issue tracker reference:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=531949

**Comment by berezovskyi** [comment]
https://git.eclipse.org/r/#/c/118582/

**Comment by berezovskyi** [comment]
@jamsden if I recall was a bit uneasy about marking RDF/JSON format & its library as deprecated. I think we should nudge all users to use JSON-LD, but I can remove the deprecation now and we can take it as a separate discussion later.

@neormx can't add you to the reviewers list but by alll means please review & provide feedback on the patch.

File: org.eclipse.lyo.oslc4j.provider.jena/pom.xml [hunk 1]
```diff
@@ -47,32 +47,58 @@
...
-                </exclusion>
-                <exclusion>
-                    <groupId>log4j</groupId>
-                    <artifactId>log4j</artifactId>
-                </exclusion>
-            </exclusions>
+            <artifactId>apache-jena-libs</artifactId>
+            <type>pom</type>
         </dependency>
+<!--      &lt;!&ndash; https://mvnrepository.com/artifact/com.fasterxml.jackson.core/jackson-core &ndash;&gt;
```
Comment (jadelkhoury) [inline org.eclipse.lyo.oslc4j.provider.jena/pom.xml]: clean up?

File: org.eclipse.lyo.oslc4j.provider.jena/src/main/java/org/eclipse/lyo/oslc4j/provider/jena/AbstractOslcRdfXmlProvider.java [hunk 2]
```diff
@@ -217,62 +242,66 @@ protected void writeTo(final boolean						queryResult,
...
-		{
-			return FileUtils.langXML;
+		List<Map.Entry<MediaType, String>> mediaPairs = new ArrayList<>();
+		mediaPairs.add(new AbstractMap.SimpleEntry<>(OslcMediaType.TEXT_TURTLE_TYPE,
+													 RDFLanguages.strLangTurtle
+		));
+		mediaPairs.add(new AbstractMap.SimpleEntry<>(OslcMediaType.APPLICATION_JSON_LD_TYPE,
+													 RDFLanguages.strLangJSONLD
+		));
+		// TODO Andrew@2018-03-03: switch to OSLC4JUtils once 118569 is merged
```
Comment (jadelkhoury) [inline org.eclipse.lyo.oslc4j.provider.jena/src/main/java/org/eclipse/lyo/oslc4j/provider/jena/AbstractOslcRdfXmlProvider.java]: I have merged that patch, you can fix this one now

File: org.eclipse.lyo.oslc4j.provider.jena/src/main/java/org/eclipse/lyo/oslc4j/provider/jena/OslcJsonLdArrayProvider.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+/**
+ * Created on 2018-03-03
+ *
+ * @author Andrew Berezovskyi (andriib@kth.se)
+ * @version $version-stub$
+ * @since 0.0.1
+ */
+@Provider
+@Produces({OslcMediaType.TEXT_TURTLE})
```
Comment (jadelkhoury) [inline org.eclipse.lyo.oslc4j.provider.jena/src/main/java/org/eclipse/lyo/oslc4j/provider/jena/OslcJsonLdArrayProvider.java]: wrong MIME

File: org.eclipse.lyo.oslc4j.provider.jena/src/main/java/org/eclipse/lyo/oslc4j/provider/jena/OslcJsonLdCollectionProvider.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+/**
+ * Created on 2018-03-03
+ *
+ * @author Andrew Berezovskyi (andriib@kth.se)
+ * @version $version-stub$
+ * @since 0.0.1
+ */
+@Provider
+@Produces({OslcMediaType.TEXT_TURTLE})
```
Comment (jadelkhoury) [inline org.eclipse.lyo.oslc4j.provider.jena/src/main/java/org/eclipse/lyo/oslc4j/provider/jena/OslcJsonLdCollectionProvider.java]: wrong MIME

File: org.eclipse.lyo.oslc4j.provider.jena/src/test/java/org/eclipse/lyo/oslc4j/provider/jena/test/JsonLdTest.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import static org.junit.Assert.*;
+
+/**
+ * Created on 2018-03-03
+ *
+ * @author Andrew Berezovskyi (andriib@kth.se)
+ * @version $version-stub$
+ * @since 0.0.1
+ */
+public class JsonLdTest {
```
Comment (jadelkhoury) [inline org.eclipse.lyo.oslc4j.provider.jena/src/test/java/org/eclipse/lyo/oslc4j/provider/jena/test/JsonLdTest.java]: test array and collection providers too?

**Comment by berezovskyi** [comment]
I have no idea why do we have Collection providers. They don't support generics and thus are virtually useless:

image

**Comment by berezovskyi** [comment]
Jad noticed that the system property access has been refactored with static methods that have a default value. He asked me to double-check that the behaviour won't change.

Here are my findings:

**Always abbreviate XML:**

The clause was `"true".equals(System.getProperty(OSLC4J_ALWAYS_XML_ABBREV))`. It means that in absence of a system property, it would evaluate to false.

**Strict datatypes:**

Cannot check right now because the refactored function is not used anywhere.

**Relative URIs:**

Was true by default https://github.com/eclipse/lyo.core/pull/3/commits/00207b45cd73549c5b7d278cbe604a56248c0506#diff-925011427795bc882b95f5a5b74429b3L416

@jamsden @jadelkhoury @yashkhatri  would be good if you could decide whether the strict datatype handling defaults are correct and then we are good to merge.

**Comment by jadelkhoury** [comment]
Concerning OSLC4J_STRICT_DATATYPES, the comments in the AbstractOslcRdfXmlProvider.java, state that the default is "true". 

Plus, the method handleExtendedPropertyValue in JenaModelHelper clearly uses that property to decide if an exception is to be thrown or not. (The logic there seems to throw an exception, unless OSLC4J_STRICT_DATATYPES is false.)

So, default to true should be fine?

**Comment by berezovskyi** [comment]
Just added the copyright headers and removed more commented out POM file stuff.

**Pull request merged**